src/Microsoft.Security.Utilities.Core/SecretMasker.cs

```diff
@@ -171,177 +174,125 @@ public string MaskSecrets(string input, Action<Detection>? detectionAction = nul
     }
 
     /// <summary>
-    /// Gets or sets the minimum allowable size of a string that's a candidate for masking.
+    /// Gets or sets the minimum length of a secret that can be detected or masked.
     /// </summary>
-    public virtual int MinimumSecretLength { get; set; }
-
-    /// <summary>
-    /// This implementation assumes no more than one thread is adding regexes, values, or encoders at any given time.
-    /// </summary>
-    [ExcludeFromCodeCoverage]
-    public void AddValue(string value)
+    public int MinimumSecretLength
     {
-        // Test for empty.
-        if (string.IsNullOrEmpty(value))
-        {
-            return;
-        }
-
-        if (value.Length < MinimumSecretLength)
-        {
-            return;
-        }
-
-        var secretLiterals = new List<SecretLiteral>(new[] { new SecretLiteral(value) });
-
-        // Read section.
-        LiteralEncoder[] literalEncoders;
-        try
+        get => _minimumSecretLength;
+        set
         {
-            SyncObject.EnterReadLock();
-
-            // Test whether already added.
-            if (ExplicitlyAddedSecretLiterals.Contains(secretLiterals[0]))
+            SyncObject.EnterWriteLock();
+            try
             {
-                return;
+                _minimumSecretLength = value;
             }
-
-            // Read the value encoders.
-            literalEncoders = LiteralEncoders.ToArray();
-        }
-        finally
-        {
-            if (SyncObject.IsReadLockHeld)
+            finally
             {
-                SyncObject.ExitReadLock();
+                SyncObject.ExitWriteLock();
             }
         }
+    }
 
-        // Compute the encoded values.
-        foreach (LiteralEncoder literalEncoder in literalEncoders)
+    public void AddValue(string value)
+    {
+        if (string.IsNullOrEmpty(value))
         {
-            string encodedValue = literalEncoder(value);
-            if (!string.IsNullOrEmpty(encodedValue) && encodedValue.Length >= MinimumSecretLength)
-            {
-                secretLiterals.Add(new SecretLiteral(encodedValue));
-            }
+            return;
         }
 
-        // Write section.
+        SyncObject.EnterWriteLock();
         try
         {
-            SyncObject.EnterWriteLock();
+            SecretLiteral literal = new SecretLiteral(value);
+            if (!_explicitlyAddedSecretLiterals.Add(literal))
```

Comment:
strictly speaking this construction is thread-safe and be completed before we take the lock.